File: strategy_runner.py
import itertools
import engine

class StrategyError(Exception):
    pass

def load_strategy(code_str):
    ns = {}
    try:
        exec(code_str, ns)
    except Exception as e:
        raise StrategyError(f"Code failed to execute: {e}")
    if "run_strategy" not in ns or not callable(ns["run_strategy"]):
        raise StrategyError("Code must define: run_strategy(df, params)")
    return ns

def parse_params(text):
    params = {}
    text = (text or "").strip()
    if not text:
        return params
    for part in text.split(","):
        if "=" not in part:
            continue
        k, v = part.split("=", 1)
        k, v = k.strip(), v.strip()
        try:
            v = float(v)
        except ValueError:
            pass
        params[k] = v
    return params

def parse_grid(text):
    grid = {}
    text = (text or "").strip()
    if not text:
        return grid
    for chunk in text.split(";"):
        if "=" not in chunk:
            continue
        k, vlist = chunk.split("=", 1)
        k = k.strip()
        vals = []
        for v in vlist.split(","):
            v = v.strip()
            try:
                vals.append(float(v))
            except ValueError:
                vals.append(v)
        grid[k] = vals
    return grid

def run_single(df, code_str, params_override=None):
    ns = load_strategy(code_str)
    params = dict(ns.get("DEFAULT_PARAMS", {}))
    params.update(params_override or {})
    trades = ns["run_strategy"](df, params)
    return {"params": params, "metrics": engine.compute_metrics(trades)}

def run_optimize(df, code_str, grid_override=None, min_trades=15):
    ns = load_strategy(code_str)
    grid = dict(ns.get("PARAM_GRID", {}))
    grid.update(grid_override or {})
    if not grid:
        raise StrategyError("No PARAM_GRID in code and no grid override given.")
    keys = list(grid.keys())
    combos = list(itertools.product(*[grid[k] for k in keys]))
    results = []
    for combo in combos:
        params = dict(zip(keys, combo))
        try:
            trades = ns["run_strategy"](df, params)
        except Exception:
            continue
        m = engine.compute_metrics(trades)
        if m["num_trades"] >= min_trades:
            results.append({"params": params, **m})
    pf_key = lambda r: (r["profit_factor"] if r["profit_factor"] != float("inf") else 9999)
    results.sort(key=pf_key, reverse=True)
    return results

def run_walkforward(df, code_str, params_override=None, split_frac=0.6):
    ns = load_strategy(code_str)
    params = dict(ns.get("DEFAULT_PARAMS", {}))
    params.update(params_override or {})
    explore, holdout, cut = engine.walk_forward_split(df, split_frac)
    ex_trades = ns["run_strategy"](explore, params)
    ho_trades = ns["run_strategy"](holdout, params)
    return {"params": params, "split_date": str(cut),
            "explore": engine.compute_metrics(ex_trades),
            "holdout": engine.compute_metrics(ho_trades)}

File: strategy_runner.js
// strategy_runner.js -- loads pasted strategy code, runs Single/Optimize/Walk-Forward.

function loadStrategy(code) {
  const mod = {};
  const wrapper = `
    ${code}
    module.runStrategy = (typeof runStrategy !== "undefined") ? runStrategy : null;
    module.DEFAULT_PARAMS = (typeof DEFAULT_PARAMS !== "undefined") ? DEFAULT_PARAMS : {};
    module.PARAM_GRID = (typeof PARAM_GRID !== "undefined") ? PARAM_GRID : {};
  `;
  try {
    const fn = new Function("module", wrapper);
    fn(mod);
  } catch (e) {
    throw new Error("Code failed to execute: " + e.message);
  }
  if (typeof mod.runStrategy !== "function") {
    throw new Error("Code must define: function runStrategy(bars, params)");
  }
  return mod;
}

function parseParams(text) {
  const params = {};
  (text || "").trim().split(",").forEach(part => {
    if (!part.includes("=")) return;
    const [k, v] = part.split("=");
    const key = k.trim(), val = v.trim();
    const num = parseFloat(val);
    params[key] = isNaN(num) ? val : num;
  });
  return params;
}

function parseGrid(text) {
  const grid = {};
  (text || "").trim().split(";").forEach(chunk => {
    if (!chunk.includes("=")) return;
    const [k, vlist] = chunk.split("=");
    const key = k.trim();
    grid[key] = vlist.split(",").map(v => {
      const num = parseFloat(v.trim());
      return isNaN(num) ? v.trim() : num;
    });
  });
  return grid;
}

function cartesianProduct(grid) {
  const keys = Object.keys(grid);
  let combos = [{}];
  for (const key of keys) {
    const next = [];
    for (const combo of combos) {
      for (const val of grid[key]) {
        next.push({ ...combo, [key]: val });
      }
    }
    combos = next;
  }
  return combos;
}

function runSingle(bars, code, paramsOverride) {
  const mod = loadStrategy(code);
  const params = { ...mod.DEFAULT_PARAMS, ...(paramsOverride || {}) };
  const trades = mod.runStrategy(bars, params);
  return { params, metrics: computeMetrics(trades), numTrades: trades.length, trades };
}

function runOptimize(bars, code, gridOverride, minTrades) {
  minTrades = minTrades || 15;
  const mod = loadStrategy(code);
  const grid = { ...mod.PARAM_GRID, ...(gridOverride || {}) };
  if (Object.keys(grid).length === 0) throw new Error("No PARAM_GRID in code and no grid override given.");
  const combos = cartesianProduct(grid);
  const results = [];
  for (const params of combos) {
    let trades;
    try { trades = mod.runStrategy(bars, params); } catch (e) { continue; }
    const m = computeMetrics(trades);
    if (m.num_trades >= minTrades) results.push({ params, ...m });
  }
  results.sort((a, b) => {
    const pfA = a.profit_factor === "inf" ? 9999 : a.profit_factor;
    const pfB = b.profit_factor === "inf" ? 9999 : b.profit_factor;
    return pfB - pfA;
  });
  return results;
}

function runWalkForward(bars, code, paramsOverride, splitFrac) {
  const mod = loadStrategy(code);
  const params = { ...mod.DEFAULT_PARAMS, ...(paramsOverride || {}) };
  const { explore, holdout, cutDate } = walkForwardSplit(bars, splitFrac);
  const exploreTrades = mod.runStrategy(explore, params);
  const holdoutTrades = mod.runStrategy(holdout, params);
  return {
    params,
    cutDate,
    explore: computeMetrics(exploreTrades),
    holdout: computeMetrics(holdoutTrades),
    exploreTrades,
    holdoutTrades,
  };
}

const DEFAULT_TEMPLATE = `// Paste/edit your strategy here. Must define runStrategy(bars, params).
// bars: array of {timestamp, date, hour, open, high, low, close} sorted ascending, UTC.
// Return: [{ date, direction: "long"/"short", r_multiple: number }, ...]

const DEFAULT_PARAMS = { session_hour: 13, sl_mult: 0.5, tp_mult: 2.0 };
const PARAM_GRID = { sl_mult: [0.3, 0.5, 0.75], tp_mult: [1.5, 2.0, 2.5] };

function runStrategy(bars, params) {
  const sessionHour = Math.round(params.session_hour !== undefined ? params.session_hour : 13);
  const slMult = params.sl_mult !== undefined ? params.sl_mult : 0.5;
  const tpMult = params.tp_mult !== undefined ? params.tp_mult : 2.0;

  const ranges = {};
  for (const b of bars) {
    if (b.hour === sessionHour) {
      if (!ranges[b.date]) ranges[b.date] = { high: b.high, low: b.low };
      else {
        ranges[b.date].high = Math.max(ranges[b.date].high, b.high);
        ranges[b.date].low = Math.min(ranges[b.date].low, b.low);
      }
    }
  }

  const trades = [];
  for (const date in ranges) {
    const rngHigh = ranges[date].high, rngLow = ranges[date].low;
    const rngSize = rngHigh - rngLow;
    if (rngSize <= 0) continue;

    const dayBars = bars.filter(b => b.date === date && b.hour > sessionHour && b.hour <= sessionHour + 2);
    let direction = null, entryPrice = null, entryPos = -1;
    for (let i = 0; i < dayBars.length; i++) {
      if (dayBars[i].close > rngHigh) { direction = "long"; entryPrice = dayBars[i].close; entryPos = i; break; }
      if (dayBars[i].close < rngLow) { direction = "short"; entryPrice = dayBars[i].close; entryPos = i; break; }
    }
    if (!direction) continue;

    let sl, tp;
    if (direction === "long") { sl = entryPrice - slMult * rngSize; tp = entryPrice + tpMult * rngSize; }
    else { sl = entryPrice + slMult * rngSize; tp = entryPrice - tpMult * rngSize; }

    const future = bars.filter(b => b.date === date && b.hour > sessionHour + 2 - 1).slice(entryPos + 1);
    let r = null;
    for (const b of future) {
      const hitSL = direction === "long" ? b.low <= sl : b.high >= sl;
      const hitTP = direction === "long" ? b.high >= tp : b.low <= tp;
      if (hitSL) { r = -1.0; break; }
      if (hitTP) { r = tpMult / slMult; break; }
    }
    if (r !== null) trades.push({ date, direction, r_multiple: r });
  }
  return trades;
}
`;
